core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,190 @@
+            create_vm(rt_handle.clone(), l1_batch_number, connection, l2_chain_id)
+                .context("failed to create vm for BasicWitnessInputProducer")?;
+
+        tracing::info!("Started execution of l1_batch: {l1_batch_number:?}");
+        let mut next_miniblocks_data = miniblocks_execution_data.iter().skip(1);
+        for miniblock_execution_data in miniblocks_execution_data.iter() {
+            tracing::debug!(
+                "Started execution of miniblock: {:?}, executing {:?} transactions",
+                miniblock_execution_data.number,
+                miniblock_execution_data.txs.len(),
+            );
+            for tx in &miniblock_execution_data.txs {
+                tracing::trace!("Started execution of tx: {tx:?}");
+                execute_tx(tx, &mut vm)
+                    .context("failed to execute transaction in BasicWitnessInputProducer")?;
+                tracing::trace!("Finished execution of tx: {tx:?}");
+            }
+            if let Some(next_miniblock_data) = next_miniblocks_data.next() {
+                vm.start_new_l2_block(L2BlockEnv::from_miniblock_data(next_miniblock_data));
+            }
+
+            tracing::debug!(
+                "Finished execution of miniblock: {:?}",
+                miniblock_execution_data.number
+            );
+        }
+        vm.finish_batch();
+        tracing::info!("Finished execution of l1_batch: {l1_batch_number:?}");
+
+        METRICS.process_batch_time.observe(started_at.elapsed());
+        tracing::info!(
+            "BasicWitnessInputProducer took {:?} for L1BatchNumber {}",
+            started_at.elapsed(),
+            l1_batch_number.0
+        );
+
+        let witness_block_state = (*storage_view).borrow().witness_block_state();
+        Ok(witness_block_state)
+    }
+}
+
+#[async_trait]
+impl JobProcessor for BasicWitnessInputProducer {
+    type Job = L1BatchNumber;
+    type JobId = L1BatchNumber;
+    type JobArtifacts = WitnessBlockState;
+    const SERVICE_NAME: &'static str = "basic_witness_input_producer";
+
+    async fn get_next_job(&self) -> anyhow::Result<Option<(Self::JobId, Self::Job)>> {
+        let mut connection = self.connection_pool.access_storage().await?;
+        let l1_batch_to_process = connection
+            .basic_witness_input_producer_dal()
+            .get_next_basic_witness_input_producer_job()
+            .await
+            .context("failed to get next basic witness input producer job")?;
+        Ok(l1_batch_to_process.map(|number| (number, number)))
+    }
+
+    async fn save_failure(&self, job_id: Self::JobId, started_at: Instant, error: String) {
+        let attempts = self
+            .connection_pool
+            .access_storage()
+            .await
+            .unwrap()
+            .basic_witness_input_producer_dal()
+            .mark_job_as_failed(job_id, started_at, error)
+            .await
+            .expect("errored whilst marking job as failed");
+        match attempts {
```

Comment:
Nit: Using `if let Some(_)` / `else` is arguably more idiomatic with `Option`s than `match`ing.